pkg/gardenlet/controller/seed/seed_care_control.go

```diff
@@ -0,0 +1,201 @@
+	if gardencorev1beta1helper.ConditionsNeedUpdate(conditions, updatedConditions) {
+		// Rebuild seed conditions and constraints to ensure that only the conditions and constraints with the
+		// correct types will be updated, and any other conditions will remain intact
+		conditions := buildSeedConditions(seed.Status.Conditions, updatedConditions, conditionTypes)
+		log.Info("Updating seed status conditions and constraints")
+		if err := patchSeedStatus(ctx, gardenClient, seed, conditions); err != nil {
+			log.Error(err, "Could not update Seed status")
+			return nil // We do not want to run in the exponential backoff for the condition checks.
+		}
+	}
+	return nil
+}
+
+func careSetupFailure(ctx context.Context, gardenClient client.Client, seed *gardencorev1beta1.Seed, message string, conditions []gardencorev1beta1.Condition) error {
+	updatedConditions := make([]gardencorev1beta1.Condition, 0, len(conditions))
+	for _, cond := range conditions {
+		updatedConditions = append(updatedConditions, gardencorev1beta1helper.UpdatedConditionUnknownErrorMessage(cond, message))
+	}
+
+	if !gardencorev1beta1helper.ConditionsNeedUpdate(conditions, updatedConditions) {
+		return nil
+	}
+
+	return patchSeedStatus(ctx, gardenClient, seed, updatedConditions)
+}
+
+// buildSeedConditions builds and returns the seed conditions using the given seed conditions as a base,
+// by first removing all conditions with the given types and then merging the given conditions (which must be of the same types).
+func buildSeedConditions(seedConditions []gardencorev1beta1.Condition, conditions []gardencorev1beta1.Condition, conditionTypes []gardencorev1beta1.ConditionType) []gardencorev1beta1.Condition {
+	result := gardencorev1beta1helper.RemoveConditions(seedConditions, conditionTypes...)
+	result = gardencorev1beta1helper.MergeConditions(result, conditions...)
+	return result
+}
+
+func patchSeedStatus(ctx context.Context, c client.StatusClient, seed *gardencorev1beta1.Seed, conditions []gardencorev1beta1.Condition) error {
+	patch := client.StrategicMergeFrom(seed.DeepCopy())
+	seed.Status.Conditions = conditions
+	return c.Status().Patch(ctx, seed, patch)
+}
+
+func (r *careReconciler) conditionThresholdsToProgressingMapping() map[gardencorev1beta1.ConditionType]time.Duration {
+	out := make(map[gardencorev1beta1.ConditionType]time.Duration)
+	for _, threshold := range r.config.Controllers.SeedCare.ConditionThresholds {
+		out[gardencorev1beta1.ConditionType(threshold.Type)] = threshold.Duration.Duration
+	}
+	return out
+}
+
+func (c *Controller) seedCareAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return
+	}
+	c.seedCareQueue.Add(key)
+}
+
+func (c *Controller) seedCareUpdate(_, newObj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(newObj)
+	if err != nil {
+		return
+	}
+
+	newSeed, ok := newObj.(*gardencorev1beta1.Seed)
+	if !ok {
+		return
+	}
+
+	if newSeed.Generation == newSeed.Status.ObservedGeneration {
```

Comment:
I think there was a misunderstanding from what @plkokanov and I meant here. `newSeed.Generation == newSeed.Status.ObservedGeneration` should actually be an indicator that there is no active reconciliation of the seed controller due to a spec change, i.e., no care operation is intended to be running at that time. So, if there is a status update && `newSeed.Generation == newSeed.Status.ObservedGeneration`, then it's fine to trigger the care operation.